### PR TITLE
Feature/ktl 4790 collect cases

### PR DESCRIPTION
**Pull request by Dan4Life**
Updated spec for malicious package-pair candidate collection

### CODE REVIEW

File: docs/specs/malicious-package-pair-candidate-collection/spec.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
+# Spec: Malicious package-pair candidate collection
+
+## 1. Goal
+Detect "malicious package pairs" — the same `artifactId` published under more than one `groupId` within a single indexed project — and persist every detected entry into a durable, reviewable table enriched with dormancy and ownership signals plus a review lifecycle (`status` + reviewer `notes`), populated by a **scheduled job that periodically recomputes the conflict set from the catalogue**, so a human can triage cases over time. This task owns both detection and collection; it makes **no ban/verdict decision**.
+
+## 2. Problem
+- There is no durable record of malicious package pairs. The detection has only ever run as a throwaway SQL view on an unmerged spike branch; nothing that ships today computes or stores these cases.
```
Comment (nikitavlaevjb) [inline docs/specs/malicious-package-pair-candidate-collection/spec.md]: The spike should not be referenced in the spec unless it is for the gathered data

File: docs/specs/malicious-package-pair-candidate-collection/spec.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+- A live view holds **no reviewer state**: a case silently appears or disappears as the catalogue changes, with no way to record "a human already looked at this."
```
Comment (nikitavlaevjb) [inline docs/specs/malicious-package-pair-candidate-collection/spec.md]: There is no live view now

File: docs/specs/malicious-package-pair-candidate-collection/spec.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+- The signals a reviewer needs aren't gathered per conflicting entry: release timestamps (dormancy) live scattered across per-version `package` rows. Nothing collects them per entry, alongside a `groupId`-derived owner login, into one triage-ready record.
```
Comment (nikitavlaevjb) [inline docs/specs/malicious-package-pair-candidate-collection/spec.md]: Thats not the problem we are solving

File: docs/specs/malicious-package-pair-candidate-collection/spec.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+- **Who's affected:** klibs.io maintainers / security reviewers triaging potential impersonation or name-squat republishes; and, downstream, the future automated ban pipeline that will consume reviewed cases.
```
Comment (nikitavlaevjb) [inline docs/specs/malicious-package-pair-candidate-collection/spec.md]: "name-squat": slop wording; please make sure that the spec contains normal human speech

File: docs/specs/malicious-package-pair-candidate-collection/spec.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+### Scenario 3 — Owner is derived only for GitHub-account coordinates (P1)
+- **Given:** two entries, one under `io.github.alice`, one under `com.example`.
+- **When:** collected.
+- **Then:** the `io.github.alice` entry row has `derived_owner_login = alice`; the `com.example` entry row has `derived_owner_login = NULL`.
+- **Independent test:** DB-integration asserting both rows' `derived_owner_login`.
+
+### Edge cases
+- Same `artifactId` under **3+ groupIds** → 3+ entry rows sharing the `(project_id, artifact_id)` group key.
+- Every candidate entry has a resolved repo: a candidate requires `project_id IS NOT NULL` (§8), and a `project` is only ever created from a resolved GitHub repo, so `project.scm_repo_id` is always present — `scm_repo_id` on the candidate is therefore non-null. (Each entry also has ≥1 release, since it derives from `package` rows whose `release_ts` is NOT NULL, so release timestamps are always present.)
```
Comment (nikitavlaevjb) [inline docs/specs/malicious-package-pair-candidate-collection/spec.md]: Is that an edge case?
Whats the point here?

File: docs/specs/malicious-package-pair-candidate-collection/spec.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+- **Why:** the detection logic has a single consumer (this collector); a persisted view would add a migration and a schema object with no other reader. Native `@Query` inside the JPA repository is the established idiom (`findAllKnownMavenCentralPackages` is a sibling aggregation; `findDuplicateDescriptions` is the same `GROUP BY … HAVING COUNT > 1` shape). The `project_id IS NOT NULL` filter is required because `package.project_id` is nullable — a package with no resolved repo has no project, and an intra-project pair requires one.
+- **Rejected:** a standalone `malicious_package_pair` view (extra DB object, only ever read by one caller).
+
+### Decision — Persist detected pairs to a table
+- **Choice:** a physical candidate table populated by the collection step, keyed on `(project_id, artifact_id, group_id)`.
+- **Why:** detection alone is stateless; a review queue needs durable rows that hold reviewer state (`status`, `notes`, `detected_at`) and survive recomputation.
+- **Rejected:** compute conflicts on demand with no persistence (nowhere to record that a human already reviewed a case).
+
+### Decision — JPA throughout (detection via a native `@Query`)
```
Comment (nikitavlaevjb) [inline docs/specs/malicious-package-pair-candidate-collection/spec.md]: That was a no-brainer here, so no need to mention it in the spec

File: docs/specs/malicious-package-pair-candidate-collection/spec.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+- **Rejected:** (a) copy `project.dependent_count` onto each row — redundant + staleness; (b) recompute a per-entry count at `group:artifact` granularity — fabricates a metric this collection task is not meant to produce.
+
+### Decision — `scm_repo_id` *is* stored on the candidate row (unlike `dependent_count`)
+- **Choice:** store `scm_repo_id` on the candidate even though it is reachable via `project`.
+- **Why:** unlike `dependent_count` (a volatile metric recomputed every 6h, so copying invites staleness), `scm_repo_id` is a **stable structural FK** — a project's repo is fixed at first index and a rename/transfer dedups back to the same `scm_repo` via GitHub `nativeId`. Storing it gives the reviewer a direct one-hop join to `scm_owner` for the later owner-mismatch check (comparing `derived_owner_login` against the real owner), without threading through `project`.
+- **Rejected:** derive it via `project` at read time — an extra hop for a value that never changes.
```
Comment (nikitavlaevjb) [inline docs/specs/malicious-package-pair-candidate-collection/spec.md]: We decided to neither derive it nor use it

File: docs/specs/malicious-package-pair-candidate-collection/spec.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+### Decision — Idempotent upsert that preserves reviewer state
```
Comment (nikitavlaevjb) [inline docs/specs/malicious-package-pair-candidate-collection/spec.md]: It was a requirement, not a decision

File: docs/specs/malicious-package-pair-candidate-collection/spec.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+- **Rejected — tail-of-drain step:** conflates 'queue drained' with 'data complete'; the drain is priority-ordered (`released_ts DESC NULLS FIRST`) and may run far past its 4h lock under a backlog, offering no real completeness guarantee.
+
+### Decision — Candidate PK type
+- **Choice:** `bigint` identity PK for the candidate table.
+- **Why:** a surrogate identity PK for a new, indefinitely-retained table (rows are kept as an audit record and never deleted — FR-011 — so the table only grows over time, even if slowly). `bigint` is the safe default, costs nothing now, and avoids any future widening migration. A generated identity is simplest for JPA. Called out because it diverges from the sibling `project` / `scm_repo` `int`/`SERIAL` PKs.
+- **Rejected:** `int`/`SERIAL` — fine at today's scale, but `bigint` future-proofs the PK for free. FK columns (`project_id`, `scm_repo_id`) stay `int` to match `project.id` / `scm_repo.id`.
+
+## 9. Key entities (only if data model changes)
+- **`MaliciousPackagePairCandidate`** (working name) — one row per entry of a conflicting pair.
```
Comment (nikitavlaevjb) [inline docs/specs/malicious-package-pair-candidate-collection/spec.md]: I'd iterate on the word "malicious".
Its time to think of something better, maybe "suspicious"? :)

File: docs/specs/malicious-package-pair-candidate-collection/spec.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+- **Detection is recompute-based:** each scheduled run recomputes the full conflict set from `package`; a pair appears on the first run after both its entries are committed. No backfill is needed — the first run seeds pre-existing conflicts. Because every run re-derives the whole truth, a missed or failed run self-heals on the next.
+- **`package.project_id` is nullable:** a package with no resolved GitHub repo has no project, so detection filters `project_id IS NOT NULL` — an intra-project pair requires a project.
+- **`package` is authoritative** for detection and timestamps: it carries `project_id`, `group_id`, `artifact_id`, `version`, `release_ts` (NOT NULL) with a unique `(group_id, artifact_id, version)` — so `version_count`, `first_release_ts`, `last_release_ts` are computable locally.
+- **Single-repo-per-project model:** both entries of a pair resolve to one repo because indexing dedups renamed/transferred repos on GitHub `nativeId`; a `project` is created only from a resolved repo, so `project.scm_repo_id` is non-null and well-defined per project (`ProjectEntity.scmRepoId: Int`).
```
Comment (nikitavlaevjb) [inline docs/specs/malicious-package-pair-candidate-collection/spec.md]: Not really related to this point, but please check:
We have multiple androidx projects, that are created manually and linked to the same scm_repo.
Would that conflict with something?

**Review by nikitavlaevjb — CHANGES_REQUESTED** [review]

File: docs/specs/malicious-package-pair-candidate-collection/spec.md (hunk 10 above)
Comment (Dan4Life) [reply docs/specs/malicious-package-pair-candidate-collection/spec.md]: I checked, it doesn't conflict right now.

**Comment by nikitavlaevjb** [comment]
If you add a new version of the spec to git as a new file, I cannot see the diff properly.
Please add changes to the previous file.

**Review by nikitavlaevjb — APPROVED** [review]

**Comment by nikitavlaevjb** [comment]
Good job, LGTM!
